CustomFeeValidator.cs

```diff
@@ -62,22 +95,43 @@ public static float GetTotalDebtForJobPurposes(DisplayableDebt debt)
         {
             if (!enabled)
                 return debt.GetTotalPrice();
-            ExistingLocoDebt locoDebt = debt as ExistingLocoDebt;
-            if (locoDebt == null)
-                return debt.GetTotalPrice();
 
-            var locoInTrainset = PlayerManager.LastLoco?.trainset?.cars?.Find(car => car.ID == locoDebt.ID);
-            if (locoInTrainset != null)
-            {
-                float fees = debt.GetTotalPriceOfResources(nonConsumableTypes);
-                DebugLog($"locomotive {locoInTrainset.ID} is in player's last trainset. Fees without consumables = {fees}");
-                return fees;
-            }
-            else
-            {
-                DebugLog($"locomotive {debt.ID} not found in last trainset");
-                return debt.GetTotalPrice();
+            switch (selectedFeeType) {
+                case FeeType.ignore_fees_from_train_with_last_loco:
+                    {
+                        ExistingLocoDebt locoDebt = debt as ExistingLocoDebt;
+                        if (locoDebt == null)
+                            return debt.GetTotalPrice();
+
+                        var locoInTrainset = PlayerManager.LastLoco?.trainset?.cars?.Find(car => car.ID == locoDebt.ID);
+                        if (locoInTrainset != null)
+                        {
+                            float fees = debt.GetTotalPriceOfResources(nonConsumableTypes);
+                            DebugLog($"Locomotive {locoInTrainset.ID} is in player's last trainset. Fees without consumables = {fees}.");
+                            return fees;
+                        }
+                        DebugLog($"Locomotive {debt.ID} not found in last trainset.");
+                        return debt.GetTotalPrice();
+                    }
+                case FeeType.ignore_all_loco_fees_until_despawned:
+                    {
+                        ExistingLocoDebt locoDebt = debt as ExistingLocoDebt;
+                        if (locoDebt == null)
+                            return debt.GetTotalPrice();
+
+                        TrainCar indebtedLoco = TrainCar.logicCarToTrainCar.Values.FirstOrDefault(tc => tc.ID == locoDebt.ID);

```

Comment:
Is this logic necessary? We already know it's an `ExistingLocoDebt`. I believe that as soon as a loco is destroyed its debts are converted to archive debts (`StagedLocoDebt`). See `LocoDebtController.StageLocoDebtOnLocoDestroy`.

This search is also a slow operation, going linearly over all the train cars in the game. It's possible to navigate from an ExistingLocoDebt to the various LocoDebtTracker subclasses and then to the damage model component of the TrainCar in constant time.

Reply:
I wasn't sure if there was a case when a destroyed loco might be represented by `ExistingLocoDebt` or not, but I'm willing to make the assumption that that will never happen. It seems like a fairly safe assumption to make.